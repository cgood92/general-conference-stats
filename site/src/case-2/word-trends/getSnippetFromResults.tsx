import React from "react";

export default function getSnippetFromResults(results: RegExpMatchArray[]) {
  return (
    <>
      ...
      {results.map((result) => {
        const { 0: textMatch, index, input } = result;

        const start = index!;
        const end = start + textMatch.length;

        const before = getBefore(input!, start);
        const mark = <mark>{textMatch}</mark>;
        const after = getAfter(input!, end);

        return (
          <React.Fragment key={index}>
            {before}
            {mark}
            {after}
          </React.Fragment>
        );
      })}
    </>
  );
}

function getBefore(input: string, index: number) {
  let cursor = 0;
  let currentCharacter = input[index - cursor];

  while (!isStopper(currentCharacter) && cursor < 80) {
    currentCharacter = input[index - ++cursor];
  }

  while (!isStopper(currentCharacter) && currentCharacter !== " ") {
    currentCharacter = input[index - --cursor];
  }

  return input.substring(index - cursor, index);
}

function getAfter(input: string, index: number) {
  let cursor = 0;
  let currentCharacter = input[index + cursor];

  while (!isStopper(currentCharacter) && cursor < 80) {
    currentCharacter = input[index + ++cursor];
  }

  while (!isStopper(currentCharacter) && currentCharacter !== " ") {
    currentCharacter = input[index + --cursor];
  }

  return input.substring(index, index + cursor) + "...";
}

const stoppers = [undefined, "\n", ".", ";"];

function isStopper(character: string) {
  return stoppers.includes(character);
}
