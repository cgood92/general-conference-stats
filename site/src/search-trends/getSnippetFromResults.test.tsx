import React from "react";
import renderer from "react-test-renderer";
import talks_1973_10 from "@root/case2/output/1973-10.json";

import getSnippetFromResults from "./getSnippetFromResults";

type Results = RegExpMatchArray[];

const Example = ({ results }: { results: Results }) =>
  getSnippetFromResults(results);

const text = talks_1973_10[0].content;

describe("creates a logical snippet for the results", () => {
  it("case 1", () => {
    const results: Results = Array.from(text.matchAll(/ment/gi));

    expect(
      renderer.create(<Example results={results} />).toJSON()
    ).toMatchSnapshot();
  });
});
