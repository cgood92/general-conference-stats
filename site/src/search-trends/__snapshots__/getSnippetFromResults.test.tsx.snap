// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a logical snippet for the results case 1 1`] = `
Array [
  "...",
  " and friends who are listening by radio and television, may I now for a few mo",
  <mark>
    ment
  </mark>,
  "s make some comments about a condition which is of great concern to all of us...",
  " listening by radio and television, may I now for a few moments make some com",
  <mark>
    ment
  </mark>,
  "s about a condition which is of great concern to all of us today...",
  "
“Self-respect is the noblest gar",
  <mark>
    ment
  </mark>,
  " with which a man can clothe himself, the most elevating feeling with which the...",
  " of the importance of conforming to certain standards required for advance",
  <mark>
    ment
  </mark>,
  "...",
  " I pondered this subject and searched the word of God to prepare for this assign",
  <mark>
    ment
  </mark>,
  ", I immediately sensed that I was to talk about a subject that is of first...",
  " often than not a bid for a kind of attention or popularity that physical and ",
  <mark>
    ment
  </mark>,
  "al endowments do not invite...",
  " not a bid for a kind of attention or popularity that physical and mental endow",
  <mark>
    ment
  </mark>,
  "s do not invite...",
  " often but a reflection of an individual who is seeking, by superficial adorn",
  <mark>
    ment
  </mark>,
  " or by abnormal conduct (in a strange way), to supply that indefinable quality...",
  " of this earth’s existence and, furthermore, if they would keep the command",
  <mark>
    ment
  </mark>,
  "s as God taught by the revelations, they would have “glory added upon their...",
  "
So the Old Testa",
  <mark>
    ment
  </mark>,
  " prophets declared with respect to death: “Then shall the dust [meaning our...",
  " blessed to have a physical body because of your obedience to certain command",
  <mark>
    ment
  </mark>,
  "s in that premortal state...",
  " apostle Paul impressed the sacredness of our individual bodies in this state",
  <mark>
    ment
  </mark>,
  ": “Know ye not that ye are the temple of God, and that the Spirit of God...",
  " members of the church that they had received the gift of a special endow",
  <mark>
    ment
  </mark>,
  " known as the Holy Ghost...",
]
`;
